filters/openpolicyagent/openpolicyagent.go

```diff
@@ -470,6 +502,100 @@ func (opa *OpenPolicyAgentInstance) MetricsKey(key string) string {
 	return key + "." + opa.bundleName
 }
 
+var (
+	ErrClosed = errors.New("reader closed")
+)
+
+type bufferedBodyReader struct {
+	input         io.ReadCloser
+	maxBufferSize uint64
+
+	bodyBuffer bytes.Buffer
+	readBuffer []byte
+
+	once   sync.Once
+	err    error
+	closed bool
+}
+
+func newBufferedBodyReader(input io.ReadCloser, maxBufferSize uint64, readBufferSize uint64) *bufferedBodyReader {
+	return &bufferedBodyReader{
+		input:         input,
+		maxBufferSize: maxBufferSize,
+		readBuffer:    make([]byte, readBufferSize),
+	}
+}
+
+func (m *bufferedBodyReader) fillBuffer(expectedSize uint64) ([]byte, error) {
+	var err error
+
+	for err == nil && uint64(m.bodyBuffer.Len()) < m.maxBufferSize && uint64(m.bodyBuffer.Len()) < expectedSize {
+		var n int
+		n, err = m.input.Read(m.readBuffer)
+
+		m.bodyBuffer.Write(m.readBuffer[:n])
+	}
+
+	if err == io.EOF {
+		err = nil
+	}
+
+	return m.bodyBuffer.Bytes(), err
+}
+
+func (m *bufferedBodyReader) Read(p []byte) (int, error) {
+	if m.closed {
+		return 0, ErrClosed
+	}
+
+	if m.err != nil {
+		return 0, m.err
+	}
+
+	// First read the buffered body
+	if m.bodyBuffer.Len() != 0 {
+		return m.bodyBuffer.Read(p)
+	}
+
+	// Continue reading from the underlying body reader
+	return m.input.Read(p)
+}
+
+// Close closes the undelrying reader if it implements io.Closer.
+func (m *bufferedBodyReader) Close() error {
+	var err error
+	m.once.Do(func() {
+		m.closed = true
+		if c, ok := m.input.(io.Closer); ok {
+			err = c.Close()
+		}
+	})
+	return err
+}
+
+func (opa *OpenPolicyAgentInstance) ExtractHttpBodyOptionally(req *http.Request) (io.ReadCloser, []byte, error) {
+	body := req.Body
+
+	if body != nil && !opa.EnvoyPluginConfig().SkipRequestBodyParse &&
+		opa.maxBodyBytes > 0 && req.ContentLength <= int64(opa.maxBodyBytes) {
```

Comment:
`opa.maxBodyBytes > 0` seems to be included in `req.ContentLength <= int64(opa.maxBodyBytes)` and 0 case should be handled by `body != nil`